File: ttycast/backends/browser.py
"""Serve the terminal at a URL and let something open it.

The lowest-common-denominator backend: no ffmpeg, no discovery, no codec
negotiation. MJPEG plays in every browser, phone and most smart-TV browsers,
and the latency is one frame.
"""

from __future__ import annotations

from ttycast.backends.base import Backend, Context
from ttycast.server import local_ip


class BrowserBackend(Backend):
    name = "browser"
    summary = "serve MJPEG at a URL; open it on the TV, a phone or a laptop"
    wants_ts = False

    def start(self, ctx: Context) -> None:
        server = ctx.server
        assert server is not None
        base = server.base_url()  # type: ignore[attr-defined]
        ctx.log(f"open on the TV:  {base}/")
        ctx.log(f"raw stream:      {base}/stream.mjpg")

    def status(self, ctx: Context) -> str:
        return f"http://{local_ip()}:{ctx.port}/"

File: src/backend/browser.rs
//! Serve the terminal at a URL and let something open it.
//!
//! The lowest-common-denominator backend: no ffmpeg, no discovery, no codec
//! negotiation. MJPEG plays in every browser, phone and most smart-TV
//! browsers, and the latency is one frame.

use super::{Backend, Context};
use crate::server::local_ip;

pub struct BrowserBackend;

impl Backend for BrowserBackend {
    fn name(&self) -> &'static str {
        "browser"
    }

    fn start(&mut self, ctx: &Context) -> Result<(), String> {
        let base = ctx
            .server
            .as_ref()
            .ok_or("the browser backend needs the http server")?
            .base_url();
        ctx.log(&format!("open on the TV:  {base}/"));
        ctx.log(&format!("raw stream:      {base}/stream.mjpg"));
        Ok(())
    }

    fn status(&self, ctx: &Context) -> String {
        format!("http://{}:{}/", local_ip(), ctx.port)
    }
}
